src/test/java/com/uber/cadence/internal/testing/WorkflowStickynessTest.java

```diff
@@ -0,0 +1,107 @@
+ *  permissions and limitations under the License.
+ */
+
+package com.uber.cadence.internal.testing;
+
+import com.uber.cadence.EventType;
+import com.uber.cadence.HistoryEvent;
+import com.uber.cadence.PollForDecisionTaskResponse;
+import com.uber.cadence.internal.testservice.TestWorkflowService;
+import com.uber.cadence.testUtils.TestServiceUtils;
+import org.junit.Test;
+
+import java.time.Duration;
+import java.util.List;
+
+import static com.uber.cadence.internal.common.InternalUtils.createNormalTaskList;
+import static com.uber.cadence.internal.common.InternalUtils.createStickyTaskList;
+import static org.junit.Assert.assertEquals;
+
+public class WorkflowStickynessTest {
+
+  @Test
+  public void taskCompletionWithStickyExecutionAttributesWillScheduleDecisionsOnStickyTaskList()
+      throws Exception {
+
+    TestWorkflowService service = new TestWorkflowService();
+    service.lockTimeSkipping("HistoryUtils");
+
+    TestServiceUtils.startWorkflowExecution("domain", "tasklist", "wfType", service);
+    PollForDecisionTaskResponse response =
+        TestServiceUtils.pollForDecisionTask("domain", createNormalTaskList("tasklist"), service);
+
+    TestServiceUtils.respondDecisionTaskCompletedWithSticky(
+        response.taskToken, "tasklist-sticky", service);
+    TestServiceUtils.signalWorkflow(response.workflowExecution, "domain", service);
+    response =
+        TestServiceUtils.pollForDecisionTask(
+            "domain", createStickyTaskList("tasklist-sticky"), service);
+
+    assertEquals(4, response.history.getEventsSize());
+    List<HistoryEvent> events = response.history.getEvents();
+    assertEquals(EventType.DecisionTaskCompleted, events.get(0).eventType);
+    assertEquals(EventType.WorkflowExecutionSignaled, events.get(1).eventType);
+    assertEquals(EventType.DecisionTaskScheduled, events.get(2).eventType);
+    assertEquals(EventType.DecisionTaskStarted, events.get(3).eventType);
+  }
+
+  @Test
+  public void taskFailureWillRescheduleTheTaskOnTheGlobalList() throws Exception {
+
+    TestWorkflowService service = new TestWorkflowService();
+    service.lockTimeSkipping("HistoryUtils");
+
+    TestServiceUtils.startWorkflowExecution("domain", "tasklist", "wfType", service);
+    PollForDecisionTaskResponse response =
+        TestServiceUtils.pollForDecisionTask("domain", createNormalTaskList("tasklist"), service);
+
+    TestServiceUtils.respondDecisionTaskCompletedWithSticky(
+        response.taskToken, "tasklist-sticky", service);
+    TestServiceUtils.signalWorkflow(response.workflowExecution, "domain", service);
+    response =
+        TestServiceUtils.pollForDecisionTask(
+            "domain", createStickyTaskList("tasklist-sticky"), service);
+    TestServiceUtils.respondDecisionTaskFailedWithSticky(response.taskToken, service);
+    response =
+        TestServiceUtils.pollForDecisionTask("domain", createStickyTaskList("tasklist"), service);
+
+    assertEquals(10, response.history.getEventsSize());
```

Comment:
This number 10 looks like a magic number to me. Maybe we just check list is non-empty and details in the last event?